errors/errors.go

```diff
@@ -125,32 +126,67 @@ type Error struct {
 }
 
 // Error returns the stored description
-func (e Error) Error() string { return e.desc }
+func (e *Error) Error() string { return e.desc }
 
 // ABCILog returns the stored description, same as Error()
-func (e Error) ABCILog() string { return e.desc }
+func (e *Error) ABCILog() string { return e.desc }
 
 // ABCICode returns the associated ABCICode
-func (e Error) ABCICode() uint32 { return e.code }
+func (e *Error) ABCICode() uint32 { return e.code }
 
 // New returns a new error. Returned instance is having the root cause set to
 // this error. Below two lines are equal
 //   e.New("my description")
 //   Wrap(e, "my description")
 // Allows sprintf format and vararg
-func (e Error) New(description string) error {
+func (e *Error) New(description string) error {
 	return Wrap(e, description)
 }
 
 // Newf is basically New with formatting capabilities
-func (e Error) Newf(description string, args ...interface{}) error {
+func (e *Error) Newf(description string, args ...interface{}) error {
 	return e.New(fmt.Sprintf(description, args...))
 }
 
-// Is is a proxy helper for global Is to be able to easily instantiate and match error codes
-// for example in tests
-func (e Error) Is(err error) bool {
-	return Is(e.New(""), err)
+// Is check if given error instance is of a given kind/type. This involves
+// unwrapping given error using the Cause method if available.
+//
+// Any non weave implementation of an error is tested positive for being
+// ErrInternal.
+func (kind *Error) Is(err error) bool {
+	type causer interface {
+		Cause() error
+	}
+
+	// Reflect usage is necessary to correctly compare with
+	// a nil implementation of an error.
+	if kind == nil {
+		if err == nil {
+			return true
+		}
+		return reflect.ValueOf(err).IsNil()
+	}
+
+	for {
+		if err == kind {
+			return true
+		}
+
+		if c, ok := err.(causer); ok {
+			err = c.Cause()
+		} else {
+
+			// As a last check, figure out if what we compare is an
+			// internal error with a non weave error.
+			if kind == ErrInternal {
```

Comment:
Hmmm... interesting.
Wonder if this is the proper use here?

I would like feedback from @alpe as well

Reply:
This is a leftover from the original design that nobody probably remembers anymore.
I find myself often referring to this as an external error as opposed to all weave errors being internal.

Meanwhile, I wonder if `ErrInternal` (or any other name) makes sense at all. For a running application, it makes sense only if we want to hide the original non-weave error.
For tests, it makes no sense to have this error, because you never want to compare non-weave errors. We only test our own code which is always returning a weave error.

Reply:
Once the Redact function is safely disabled in all tests #363 
We can then Redact Internal Errors as well at the abci interface

Reply:
For equality we should be more precise.`ErrInternal` is not the same as any random stdlib error. Let's return false in this case. I agree that this case should only apply to tests.

Reply:
Changed in https://github.com/iov-one/weave/pull/393/commits/1c5a1028c7c78ee515e4808d37212f14fac8f3d0